lib/kernel/src/application_controller.erl

```diff
@@ -372,42 +372,47 @@ get_pid_key(Master, Key) ->
 	_ -> undefined
     end.
 
-get_key(AppName, Key) ->
-    case ets:lookup(ac_tab, {loaded, AppName}) of
-	[{_, Appl}] ->
-	    case Key of 
-		description ->
-		    {ok, Appl#appl.descr};
-		id ->
-		    {ok, Appl#appl.id};
-		vsn ->
-		    {ok, Appl#appl.vsn};
-		modules ->
-		    {ok, (Appl#appl.appl_data)#appl_data.mods};
-		maxP ->
-		    {ok, (Appl#appl.appl_data)#appl_data.maxP};
-		maxT ->
-		    {ok, (Appl#appl.appl_data)#appl_data.maxT};
-		registered ->
-		    {ok, (Appl#appl.appl_data)#appl_data.regs};
-		included_applications ->
-		    {ok, Appl#appl.inc_apps};
-                optional_applications ->
-                    {ok, Appl#appl.opt_apps};
-		applications ->
-		    {ok, Appl#appl.apps};
-		env ->
-		    {ok, get_all_env(AppName)};
-		mod ->
-		    {ok, (Appl#appl.appl_data)#appl_data.mod};
-		start_phases ->
-		    {ok, (Appl#appl.appl_data)#appl_data.phases};
-		_ -> undefined
-	    end;
-	_ ->
-	    undefined
-    end.
-	    
+get_key(AppName, description) ->
+	get_key_direct(AppName, #appl{descr = '$1', _ = '_'});
+get_key(AppName, id) ->
+	get_key_direct(AppName, #appl{id = '$1', _ = '_'});
+get_key(AppName, vsn) ->
+	get_key_direct(AppName, #appl{vsn = '$1', _ = '_'});
+get_key(AppName, modules) ->
+	get_key_data(AppName, #appl_data{mods = '$1', _ = '_'});
+get_key(AppName, maxP) ->
+	get_key_data(AppName, #appl_data{maxP = '$1', _ = '_'});
+get_key(AppName, maxT) ->
+	get_key_data(AppName, #appl_data{maxT = '$1', _ = '_'});
+get_key(AppName, registered) ->
+	get_key_data(AppName, #appl_data{regs = '$1', _ = '_'});
+get_key(AppName, included_applications) ->
+	get_key_direct(AppName, #appl{inc_apps = '$1', _ = '_'});
+get_key(AppName, optional_applications) ->
+	get_key_direct(AppName, #appl{opt_apps = '$1', _ = '_'});
+get_key(AppName, applications) ->
+	get_key_direct(AppName, #appl{apps = '$1', _ = '_'});
+get_key(AppName, env) ->
+	case ets:member(ac_tab, {loaded, AppName}) of
+		true -> {ok, get_all_env(AppName)};
+		false -> undefined
+	end;
+get_key(AppName, mod) ->
+	get_key_data(AppName, #appl_data{mod = '$1', _ = '_'});
+get_key(AppName, start_phases) ->
+	get_key_data(AppName, #appl_data{phases = '$1', _ = '_'});
+get_key(_, _) ->
+	undefined.
+
+get_key_direct(AppName, Match) ->
+	case ets:match(ac_tab, {{loaded, AppName}, Match}) of
```

Comment:
All of the matches appear to be looking for a single field of the records. Is it possible to use ets:lookup_element/4 rather than ets:match/3 ?
Rather than creating and passing a pattern, it seems like the code could just pass a field index (e.g. `#appl-data.maxT` rather than `#appl_data{maxT = '$1', _ = '_'}`.
I don't know whether it would be significantly faster but it seems like it would be easy to measure.

LGTM apart from that suggestion.

Reply:
Unfortunately lookup_element wouldn't work since this isn't a top-level tuple - the values are nested in a tuple that's a field of the top-level one. lookup_element can only select pieces of the top-level tuple